store/localstore/local_region.go

```diff
@@ -44,18 +46,133 @@ type regionResponse struct {
 
 const chunkSize = 64
 
+type sortRow struct {
+	key  []types.Datum
+	meta tipb.RowMeta
+	data []byte
+}
+
+type topnContainer struct {
+	orderByItems []*tipb.ByItem
+	rows         []*sortRow
+	err          error
+}
+
+func (t *topnContainer) Len() int {
+	return len(t.rows)
+}
+
+func (t *topnContainer) Swap(i, j int) {
+	t.rows[i], t.rows[j] = t.rows[j], t.rows[i]
+}
+
+func (t *topnContainer) Less(i, j int) bool {
+	for index, by := range t.orderByItems {
+		v1 := t.rows[i].key[index]
+		v2 := t.rows[j].key[index]
+
+		ret, err := v1.CompareDatum(v2)
+		if err != nil {
+			t.err = errors.Trace(err)
+			return true
+		}
+
+		if by.Desc {
+			ret = -ret
+		}
+
+		if ret < 0 {
+			return true
+		} else if ret > 0 {
+			return false
+		}
+	}
+
+	return false
+}
+
+type topnSolver struct {
+	topnContainer
+
+	// totalCount is equal to the limit count, which means the max size of heap.
+	totalCount int
+	// heapSize means the current size of this heap.
+	heapSize int
+}
+
+func (t *topnSolver) Len() int {
+	return t.heapSize
+}
+
+func (t *topnSolver) Push(x interface{}) {
+	t.rows = append(t.rows, x.(*sortRow))
+	t.heapSize++
+}
+
+func (t *topnSolver) Pop() interface{} {
+	return nil
+}
+
+func (t *topnSolver) Less(i, j int) bool {
+	for index, by := range t.orderByItems {
+		v1 := t.rows[i].key[index]
+		v2 := t.rows[j].key[index]
+
+		ret, err := v1.CompareDatum(v2)
+		if err != nil {
+			t.err = errors.Trace(err)
+			return true
+		}
+
+		if by.Desc {
+			ret = -ret
+		}
+
+		if ret > 0 {
+			return true
+		} else if ret < 0 {
+			return false
+		}
+	}
+
+	return false
+}
+
+func (t *topnSolver) tryToAddRow(row *sortRow) bool {
+	success := false
+	if t.heapSize == t.totalCount {
+		t.rows = append(t.rows, row)
+		if t.Less(0, t.heapSize) {
+			t.Swap(0, t.heapSize)
+			heap.Fix(t, 0)
+			success = true
+		}
+		t.rows = t.rows[:t.heapSize]
+	} else {
+		heap.Push(t, row)
+		success = true
+	}
+	return success
+}
+
 type selectContext struct {
 	sel          *tipb.SelectRequest
 	txn          kv.Transaction
 	eval         *xeval.Evaluator
 	whereColumns map[int64]*tipb.ColumnInfo
 	aggColumns   map[int64]*tipb.ColumnInfo
+	topnColumns  map[int64]*tipb.ColumnInfo
 	groups       map[string]bool
 	groupKeys    [][]byte
 	aggregates   []*aggregateFuncExpr
-	aggregate    bool
+	topnSolver   *topnSolver
 	keyRanges    []kv.KeyRange
 
+	// TODO: Only one of these three flags can be true at the same time. We should set this as an enum var.
+	aggregate bool
+	desc      bool
```

Comment:
At the first sight, we would think `desc` and other two flags can be true at the same time.
Maybe `desc` should be renamed to something more specific?